### DISCUSSION TITLE
GPIO pinmux

### DISCUSSION THREAD

**Discussion by biot** [mcknly/breadboard-os#5] [Ideas]
I think BBOS should have an internal model (hardware-specific) of available GPIOs and other functions (SPI, I2C) available behind them via pinmux stuff. Currently the pins are hardcoded to one function or other, so e.g. there's no way to tell BBOS at compile time (via project.cmap) or runtime that you connected your BME280 on I2C1 instead of I2C0.

I think the pinmix configuration that is compiled in should only be the default at bootup, but switchable at runtime, like this: (this is fake, no code yet)
```
[pico /]$ pinouts enable uart1
gpio 6 switched from GP4 to UART1_TX
gpio 7 switched from GP5 to UART1_RX
[pico /]$
```

The `pinouts` command is a generic command that lives in `cli/`, with a hardware-specific backend under `hardware/` that has the list of available pins and mappings of what they can mux to, as well as available peripherals. That means it can also trivially report on what's available: (this is from working code)

```
[pico /]$ pinouts
gpio  1: GP0  SPI0_RX    I2C0_SDA   UART0_TX  
gpio  2: GP1  SPI0_CSn   I2C0_SCL   UART0_RX  
gpio  4: GP2  SPI0_SCK   I2C1_SDA  
gpio  5: GP3  SPI0_TX    I2C1_SCL  
gpio  6: GP4  SPI0_RX    I2C0_SDA   UART1_TX  
[pico /]$  
```

Once in place, this will trivially allow things like this: (fake again)
```
[pico /]$ console add uart1
[pico /]$ driver enable bme280 port i2c1
```

Opinions?

**Comment by mcknly** [comment]
Interesting idea. Definitely makes BBOS more of a "swiss army knife" bench tool for breadboarding that could be used out of the box for prototyping without having to write or modify code. Could be quite useful for a wider user base too. Your 'driver' command could also list the available drivers that have been compiled in, since we were discussing making everything more modular/optional.

I briefly considered implementing something like your 'pinouts' command that would print out an ASCII image of the board being used with pin map labels, I'm sure you have noticed my love of ASCII art...

**Comment by biot** [comment]
You know this `pinouts` output is good, but it doesn't preclude another command that uses this same information to draw a pinout diagram like [this](http://land-boards.com/blwiki/images/thumb/5/56/Raspberry-Pi-Pico-Pinout.jpg/1460px-Raspberry-Pi-Pico-Pinout.jpg). I'll add a board-specific mapping from MCU pin numbers to board silkscreen markings. I have an adafruit rp2040 board, and it's already very different from the pico in that regard.
